Rust/src/common/divector.rs

```diff
@@ -0,0 +1,174 @@
+use crate::samplerplustree::boundingbox::BoundingBox;
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct DiVector {
+    pub high: Vec<f64>,
+    pub low: Vec<f64>,
+}
+
+impl DiVector {
+    pub fn empty(dimension: usize)  -> Self {
+        DiVector{
+            high: vec![0.0;dimension],
+            low: vec![0.0;dimension]
+        }
+    }
+
+    pub fn new(high : &[f64], low :&[f64]) -> Self{
+        assert!(high.len() == low.len(), " incorrect lengths");
+        DiVector{
+            high : Vec::from(high),
+            low : Vec::from(low)
+        }
+    }
+
+    pub fn assign_as_probability_of_cut(&mut self, bounding_box: &BoundingBox, point: &[f32]) {
+        let minsum: f64 = self.low
+            .iter_mut()
+            .zip(bounding_box.get_min_values())
+            .zip(point)
+            .map(|((x, &y), &z)| if y - z > 0.0 {
+                *x = (y - z) as f64;
+                *x
+            } else {
+                *x = 0.0;
+                *x
+            })
+            .sum();
+        let maxsum: f64 = self.high
+            .iter_mut()
+            .zip(point)
+            .zip(bounding_box.get_max_values())
+            .map(|((x, &y), &z)| if y - z > 0.0 {
+                *x = (y - z) as f64;
+                *x
+            } else {
+                *x = 0.0;
+                *x
+            })
+            .sum();
+
+        let sum = minsum + maxsum;
+        if sum != 0.0 {
+            self.scale(1.0/(bounding_box.get_range_sum() + sum));
+        }
+    }
+
+    pub fn assign_as_probability_of_cut_with_missing_coordinates(
+        &mut self,
+        bounding_box:&BoundingBox,
+        point: &[f32],
+        missing_coordinates: &[bool]
+    ) {
+        let minsum: f64 = self.low
+            .iter_mut()
+            .zip(bounding_box.get_min_values())
+            .zip(point)
+            .zip(missing_coordinates)
+            .map(|(((x, &y), &z), &b)| if !b && y - z > 0.0 {
+                *x = (y - z) as f64;
+                *x
+            } else {
+                *x = 0.0;
+                *x
+            })
+            .sum();
+        let maxsum: f64 = self.high
+            .iter_mut()
```

Comment:
At some point we should do an experiment to see if there's any performance penalty to using all these `zip`s together versus a for-loop and accessing values by index.

Reply:
Its a 50%+ penalty to not have these iterators